yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -372,7 +372,14 @@ private[spark] class ApplicationMaster(
             case i: InterruptedException =>
             case e: Throwable => {
               failureCount += 1
-              if (!NonFatal(e) || failureCount >= reporterMaxFailures) {
+              if ("org.apache.hadoop.yarn.exceptions.ApplicationAttemptNotFoundException".equals(
+                e.getClass().getName())) {
+                val message = "ApplicationAttemptNotFoundException was thrown " +
+                  "from Reporter thread."
+                logError(message, e)
+                finish(FinalApplicationStatus.FAILED, ApplicationMaster.EXIT_REPORTER_FAILURE,
+                  message)
+              } else if (!NonFatal(e) || failureCount >= reporterMaxFailures) {
```

Comment:
I'm pretty confused about this: @tgravescs @vanzin shouldn't we do this only for `NonFatal` exceptions? For fatal ones we shouldn't even bother catching it.